### PR TITLE
Merkle Proof refactor

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
Added a new endpoint that returns the proof for an account, and the proof for a given key from the data trie, all in a single call.

The following endpoints need to be tested: 
"/proof/address/:address"
"/proof/root-hash/:roothash/address/:address"
"/proof/root-hash/:roothash/address/:address/key/:key"
"/proof/verify"

### CODE REVIEW

File: api/groups/baseProofGroup_test.go [hunk 1]
```diff
@@ -172,14 +168,89 @@ func TestVerifyProof(t *testing.T) {
...
+			assert.Equal(t, key, k)
+			return &data.GenericAPIResponse{Data: proof}, nil
+		},
+	}
+
+	proofGroup, err := groups.NewProofGroup(facade)
+	require.NoError(t, err)
+	ws := startProxyServer(proofGroup, "/proof")
+
+	endpoint := "/proof/root-hash/" + rootHash + "/address/" + address + "/key/" + key
```
Comment (bogdan-rosianu) [inline api/groups/baseProofGroup_test.go:236]: optional: could have used fmt.Sprintf

File: api/mock/facadeStub.go [hunk 2]
```diff
@@ -8,8 +8,8 @@ import (
 	"github.com/ElrondNetwork/elrond-proxy-go/data"
 )
 
-// Facade is the mock implementation of a node's router handler
-type Facade struct {
+// FacadeStub is the mock implementation of a node's router handler
```
Comment (bogdan-rosianu) [inline api/mock/facadeStub.go:11]: `// FacadeStub is the stub implementation of the proxy routes' handler`

File: cmd/proxy/config/apiConfig/v1_0.toml [hunk 3]
```diff
@@ -97,6 +97,7 @@ Routes = [
 [APIPackages.proof]
 Routes = [
     { Name = "/root-hash/:roothash/address/:address", Secured = false, Open = false, RateLimit = 0 },
+    { Name = "/root-hash/:roothash/address/:address/key/:key", Secured = false, Open = false, RateLimit = 0 },
```
Comment (bogdan-rosianu) [inline cmd/proxy/config/apiConfig/v1_0.toml:100]: so by default you want these endpoints to be closed, right?

Comment (iulianpascalau) [reply cmd/proxy/config/apiConfig/v1_0.toml:100]: we should have them opened for internal testnets but closed for the mainnet/testnet/devnet versions

File: process/proofProcessor.go [hunk 4]
```diff
@@ -71,6 +71,49 @@ func (pp *ProofProcessor) GetProof(rootHash string, address string) (*data.Gener
...
 }
 
+// GetProofDataTrie sends the request to the right observer and then replies with the returned answer
+func (pp *ProofProcessor) GetProofDataTrie(rootHash string, address string, key string) (*data.GenericAPIResponse, error) {
+	observers, err := pp.getObserversForAddress(address)
+	if err != nil {
+		return nil, err
+	}
+
+	getProofDataTrieEndpoint := "/proof/root-hash/" + rootHash + "/address/" + address + "/key/" + key
```
Comment (iulianpascalau) [inline process/proofProcessor.go:81]: fmt.Sprintf

File: process/proofProcessor.go [hunk 5]
```diff
@@ -71,6 +71,49 @@ func (pp *ProofProcessor) GetProof(rootHash string, address string) (*data.Gener
 	return nil, ErrSendingRequest
 }
 
+// GetProofDataTrie sends the request to the right observer and then replies with the returned answer
+func (pp *ProofProcessor) GetProofDataTrie(rootHash string, address string, key string) (*data.GenericAPIResponse, error) {
```
Comment (iulianpascalau) [inline process/proofProcessor.go:75]: unit test?

File: api/groups/baseProofGroup_test.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply api/groups/baseProofGroup_test.go:236]: Done.

File: api/mock/facadeStub.go (hunk 2 above)
Comment (BeniaminDrasovean) [reply api/mock/facadeStub.go:11]: Done.

File: cmd/proxy/config/apiConfig/v1_0.toml (hunk 3 above)
Comment (BeniaminDrasovean) [reply cmd/proxy/config/apiConfig/v1_0.toml:100]: Changed to open. They will be closed on mainnet/testnet/devnet.

File: process/proofProcessor.go (hunk 5 above)
Comment (BeniaminDrasovean) [reply process/proofProcessor.go:75]: Added.

File: cmd/proxy/config/apiConfig/v1_0.toml [hunk 6]
```diff
@@ -96,8 +96,8 @@ Routes = [
 
 [APIPackages.proof]
 Routes = [
-    { Name = "/root-hash/:roothash/address/:address", Secured = false, Open = false, RateLimit = 0 },
-    { Name = "/root-hash/:roothash/address/:address/key/:key", Secured = false, Open = false, RateLimit = 0 },
-    { Name = "/address/:address", Secured = false, Open = false, RateLimit = 0 },
-    { Name = "/verify", Secured = false, Open = false, RateLimit = 0 }
+    { Name = "/root-hash/:roothash/address/:address", Secured = false, Open = true, RateLimit = 0 },
```
Comment (iulianpascalau) [inline cmd/proxy/config/apiConfig/v1_0.toml:99]: we should revert these to `Open = false`

File: cmd/proxy/config/apiConfig/v_next.toml [hunk 7]
```diff
@@ -96,8 +96,8 @@ Routes = [
 
 [APIPackages.proof]
 Routes = [
-    { Name = "/root-hash/:roothash/address/:address", Secured = false, Open = false, RateLimit = 0 },
-    { Name = "/root-hash/:roothash/address/:address/key/:key", Secured = false, Open = false, RateLimit = 0 },
-    { Name = "/address/:address", Secured = false, Open = false, RateLimit = 0 },
-    { Name = "/verify", Secured = false, Open = false, RateLimit = 0 }
+    { Name = "/root-hash/:roothash/address/:address", Secured = false, Open = true, RateLimit = 0 },
```
Comment (iulianpascalau) [inline cmd/proxy/config/apiConfig/v_next.toml:99]: we should revert these to `Open = false`

**Review by bogdan-rosianu — DISMISSED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**